snow/engine/common/timer.go

```diff
@@ -3,12 +3,87 @@
 
 package common
 
-import "time"
+import (
+	"sync"
+	"time"
+)
 
-// Timer describes the standard interface for specifying a timeout
-type Timer interface {
+// PreemptionSignal signals when to preempt the timeouts of the timeout handler.
+type PreemptionSignal struct {
+	activateOnce sync.Once
+	initOnce     sync.Once
+	signal       chan struct{}
+}
+
+func (ps *PreemptionSignal) init() {
+	ps.signal = make(chan struct{})
+}
+
+// Listen returns a read-only channel that is closed when Preempt() is invoked.
+func (ps *PreemptionSignal) Listen() <-chan struct{} {
+	ps.initOnce.Do(ps.init)
+	return ps.signal
+}
+
+// Preempt causes any past and future call of Listen return a closed channel.
+func (ps *PreemptionSignal) Preempt() {
+	ps.initOnce.Do(ps.init)
+	ps.activateOnce.Do(func() {
+		close(ps.signal)
+	})
+}
+
+type timeoutHandler struct {
+	newTimer         func(time.Duration) (<-chan time.Time, func() bool)
+	timeouts         chan struct{}
+	onTimeout        func()
+	preemptionSignal <-chan struct{}
+}
+
+// NewTimeoutHandler constructs a new timeout handler with the given function to be invoked upon a timeout,
+// unless the preemptionSignal is closed and in which case it invokes the function immediately.
+func NewTimeoutHandler(onTimeout func(), preemptionSignal <-chan struct{}) *timeoutHandler {
+	timeoutBuff := make(chan struct{}, 1)
+	return &timeoutHandler{
+		preemptionSignal: preemptionSignal,
+		newTimer: func(d time.Duration) (<-chan time.Time, func() bool) {
+			timer := time.NewTimer(d)
+			return timer.C, timer.Stop
+		},
```

Comment:
What's the reason for specifying this function here rather than just using `time.NewTimer` directly?

It looks like we replace this with an identical implementation in one of our test cases

Reply:
Because I needed it for the tests :-)